clients/drcachesim/simulator/snoop_filter.h

```diff
@@ -0,0 +1,36 @@
+
+#ifndef _SNOOP_FILTER_H_
+#define _SNOOP_FILTER_H_ 1
+
+#include "cache.h"
+#include <unordered_map>
+#include <vector>
+
+struct coherence_table_entry_t {
+    std::vector<bool> sharers;
+    bool dirty;
+};
+
+class snoop_filter_t {
+public:
+    snoop_filter_t(void);
+    virtual bool
+    init(cache_t **caches_, int num_snooped_caches_);
+    virtual void
+    snoop(addr_t tag, int id_in, bool is_write);
+    virtual void
+    snoop_eviction(addr_t tag, int id_in);
+    void
+    print_stats(void);
+
+protected:
+    // XXX: This initial coherence implementation uses a perfect snoop filter.
+    std::unordered_map<addr_t, coherence_table_entry_t> coherence_table;
+    cache_t **caches;
+    int num_snooped_caches;
+    int_least64_t num_writes;
+    int_least64_t num_writebacks;
+    int_least64_t num_invalidates;
+};
+
+#endif /* _SNOOP_FILTER_H_ */
```

Comment:
Please add some tests.  At least add one end-to-end test that turns the option on so the code is exercised.  Possibilities:

- Simple unit tests in drcachesim_unit_tests for certain staged sequences of events
- End-to-end online test: though w/ non-deterministic results, to avoid flakiness its golden output has to be quite fuzzy and it really mostly tests that nothing crashes. Sthg like `torunonly_drcachesim(coherence client.annotation-concurrency "-coherence" "")`.
- For a stricter numeric result end-to-end test, you could check in a small offline trace and process that (see the offline-legacy or offline-snappy tests for examples)

Reply:
I hadn't used the test suite before, but I'll get familiar with it and add a test soon.

Reply:
I've added the coherence test and it passes. I believe the latest AppVeyor run failed a test because of the issue documented here:
https://github.com/DynamoRIO/dynamorio/issues/3320

Reply:
Thanks.  I'm out today but will try to look by mid-week if nobody else has reviewed by then.